### PR TITLE
Stable cache keys using DB session ids in SSTs

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Use DB session ids in SST table properties to make cache keys
stable across DB re-open and copy / move / restore / etc.

These new cache keys are currently only enabled when FileSystem does not
provide GetUniqueId. For now, they are typically larger, so slightly
less efficient.

Relevant to #7405

This change has a minor regression in PersistentCache functionality:
metaindex blocks are no longer cached in PersistentCache. Table properties
blocks already were not but ideally should be. I didn't spent effort to
fix & test these issues because we don't believe PersistentCache is used much
if at all and expect SecondaryCache to replace it. (Though PRs are welcome.)

FIXME: there is more to be fixed for stable cache keys on external SST files

Test Plan: new unit test added, which fails when disabling new
functionality

### CODE REVIEW

File: table/block_based/block_based_table_reader.cc
Comment (zhichao-cao) [inline table/block_based/block_based_table_reader.cc:390]: I guess, no one is using the persistent cache. Maybe we can deperate it in the future when secondary cache is ready

Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:390]: Microsoft might be using it. See #6932 @zhli1142015

File: table/block_based/block_based_table_reader.cc
Comment (zhichao-cao) [inline table/block_based/block_based_table_reader.cc:649]: Maybe also mention that cc->NewId() will be used instead file_number as the suffix.

**Review by zhichao-cao — APPROVED** [review]
LGTM, thanks for the PR, it will make the db_session_id stable when db is reopened.

File: table/block_based/block_based_table_reader.cc
Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:649]: Oops! My intention was to use a fixed file number as suffix but 0 means "unknown."

I will fix this and add to unit test.

Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:649]: Actually, fixing it requires a bit more intervention than I want to throw into this PR after approval.

Comment (zhichao-cao) [reply table/block_based/block_based_table_reader.cc]: I will take a look after you push the updates.

**Pull request closed without merging**
